### PR TITLE
Refactor UploadedImagesComponent

### PR DESCRIPTION
**Pull request by davidgisbey**
## Description

At the moment, the UploadedImagesComponent does a lot. It handles rendering the lead image and associated guidance, in addition to all the images that can be used inline within the body.

This adds two additional components which handle rendering the lead image and other images. I've added a bunch of unit tests for both. There's definitely some duplication, particularly around displaying the image itself, but overall i think it's the lesser evil.

## Trello card

https://trello.com/c/5vCiIxeQ/1064-lead-image-handling-backend-work

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

### CODE REVIEW

File: app/components/admin/edition_images/lead_image_component.html.erb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        <div class="govuk-grid-column-two-thirds">
+          <p class="govuk-body"><strong>Caption: </strong><%= caption %></p>
+          <p class="govuk-body"><strong>Alt text: </strong><%= alt_text %></p>
+        </div>
+      </div>
+    <% end %>
+
+    <div class="app-view-edition-resource__actions govuk-grid-column-full govuk-button-group">
+      <% if case_study? %>
+        <%= form_with(url: update_image_display_option_admin_edition_path(edition), method: :patch) do |form| %>
```
Comment (davidgisbey) [inline app/components/admin/edition_images/lead_image_component.html.erb:33]: previously this was wrapped in the form so there was a bit less duplication. It feels wrong to render a form that can't be submitted/isn't relevant to the edition type though. Interested to see what people think.

**Review by ryanb-gds — APPROVED** [review]
Couple of suggestions, one of which is probably absolute rubbish, but the other one seems like a small but easy improvement. Looks good though

File: app/components/admin/edition_images/image_component.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    edition.images_have_unique_filenames? ? "[Image: #{image.filename}]" : "!!#{find_index_from_non_lead_images + 1}"
+  end
+
+  def find_index_from_non_lead_images
+    if edition.respond_to?(:non_lead_images)
+      edition.non_lead_images.find_index(image)
+    else
+      edition.images.find_index(image)
+    end
+  end
```
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/components/admin/edition_images/image_component.rb:36]: Yea if we had access to the document_images method then we could just use that here. 

We could potentially inherit from the uploaded images component. Or we could pass the document_images in as an additional parameter too.

[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/components/admin/edition_images/image_component.rb:32]: what do you think about doing this https://github.com/alphagov/whitehall/pull/8440/commits/ead27d5a814055d568bf9579bde1022215170928?

[the remark this answers is not in the document]
Comment (ryanb-gds) [reply app/components/admin/edition_images/image_component.rb:32]: I'm not keen on that, those components are not really an extension of each other. Could we get the index before we render the component somehow?

[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/components/admin/edition_images/image_component.rb]: We could do something like [this](https://github.com/alphagov/whitehall/pull/8440/commits/b42efbcb48bab66909e0d0360d7ac790b0bda5aa) instead.  

We still need an additional method which returns the index of the image with the lead image included for the markdown index number though.

File: app/components/admin/edition_images/lead_image_component.html.erb (hunk 1 above)
Comment (ryanb-gds) [reply app/components/admin/edition_images/lead_image_component.html.erb:33]: Do the links need to be inside the form when it is a case study? It looks like they should always render when the lead image is present, so could they just go outside of the case study condition? Appreciate we might have to tweak some styles

File: app/components/admin/edition_images/image_component.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryanb-gds) [reply app/components/admin/edition_images/image_component.rb]: I think how it is at the moment looks better. In an ideal world perhaps the `ImageComponent` would not need to have access to the edition, but I don't see it as a major problem and I don't think that's very realistic.

**Pull request merged**